### PR TITLE
net: conn_mgr: Cleanup & Connectivity Counting

### PR DESCRIPTION
**Pull request by glarsennordic**
Key features:
- moved from subsys/net/lib/conn_mgr to just subsys/net/conn_mgr
- event logic and flag names simplified
- thread-safety added
- basic filtering of ifaces and L2s added
- NET_EVENT_L4_CONNECTED/NET_EVENT_L4_DISCONNECTED now only fired when the first iface gains connectivity, and when the last iface loses connectivity, respectively

### CODE REVIEW

**Comment by glarsennordic** [comment]
Currently a draft:

- Based on https://github.com/zephyrproject-rtos/zephyr/pull/55946, so ignore the first two commits
- Supercedes https://github.com/zephyrproject-rtos/zephyr/pull/54838
- Currently missing a test suite

File: tests/net/conn_mgr/src/main.c
Comment (glarsennordic) [inline tests/net/conn_mgr/src/main.c:484]: Thinking on it, this suite currently lacks a test explicitly requiring DAD to succeed as part of the test. I will write one.

File: tests/net/conn_mgr/src/main.c
Comment (glarsennordic) [inline tests/net/conn_mgr/src/main.c:522]: DAD sleep no longer makes sense for this specific routine

File: include/zephyr/net/conn_mgr.h
Comment (wbober) [inline include/zephyr/net/conn_mgr.h:30]: I think `*_watch` and `*_unwatch` could also work.

Should we explain what is the implication of ignoring an interface or it's just too obvious?

File: include/zephyr/net/conn_mgr.h
Comment (wbober) [inline include/zephyr/net/conn_mgr.h:23]: From the API user perspective, what is the purpose of that function?

File: samples/net/sockets/echo_server/src/tunnel.c
Comment (wbober) [inline samples/net/sockets/echo_server/src/tunnel.c:171]: I find this part a bit unfortunate. This is because you need to explicitly know to add this for a given interface type. How do you, as a developer, will know that it's needed?

Perhaps a method with an interface or L2 flag, which indicates if an interface is suitable for watching by `conn_mgr`, would make sense. This could be added in a subsequent PR. This way we shift the responsibility from the developer to integration author.

**Review by wbober — DISMISSED** [review]

File: subsys/net/conn_mgr/conn_mgr.c
Comment (rlubos) [inline subsys/net/conn_mgr/conn_mgr.c]: "Also rename a few flags." no longer seems to apply to commit 003a420e804be12170241674ab8a351f2de42492?

File: subsys/net/conn_mgr/conn_mgr.c
Comment (rlubos) [inline subsys/net/conn_mgr/conn_mgr.c]: nit: This commit could already use proper flags, instead of having them fixed in the next one.

File: include/zephyr/net/conn_mgr.h
Comment (rlubos) [inline include/zephyr/net/conn_mgr.h:44]: If you document individual return values, `@retval` should be used instead.

File: include/zephyr/net/conn_mgr.h
Comment (rlubos) [reply include/zephyr/net/conn_mgr.h:30]: I don't really have a strong preference, but `net_conn_mgr_monitor_iface` could also be an option (with an extra bool parameter to enable/disable).

File: samples/net/sockets/echo_server/src/echo-server.c
Comment (rlubos) [inline samples/net/sockets/echo_server/src/echo-server.c:87]: The changes in `echo_server` sample seem a bit unrelated to other work in this commit, perhaps they merit a separate commit?

File: include/zephyr/net/conn_mgr.h
Comment (glarsennordic) [reply include/zephyr/net/conn_mgr.h:30]: I don't have a strong preference on naming either, though I lean towards "ignore/unignore" simply because it suggests that ifaces are paid attention by default (which they are). @wbober Since Robert doesn't have a strong opinion, and you raised the issue originally, which of the three options would be your final choice?

I agree that the documentation for these functions needs to be improved. I will make sure to do that

File: include/zephyr/net/conn_mgr.h
Comment (glarsennordic) [reply include/zephyr/net/conn_mgr.h:23]: It is a legacy feature from the previous version of `conn_mgr`. It appears to be provided so that applications that set up their event listeners after conn_mgr initialized can manually request the latest connectivity state be resent.

I think it might be worth it to eventually phase this function out, but I consider that task beyond the scope of these modifications

File: samples/net/sockets/echo_server/src/tunnel.c
Comment (glarsennordic) [reply samples/net/sockets/echo_server/src/tunnel.c:171]: I'm not sure the responsibility should be with the integration provider; I can imagine a scenario where, for instance, where a connectivity-capable iface might be occasionally available, but the application developer does not want to rely on it. I also think that whether an iface is suitable for monitoring depends largely on context, and that context is defined in the application domain.

But I think I agree with you partially, that `net_conn_mgr_ignore_l2` is not really a great way for applications to control which ifaces they receive. The L2 type is often very disconnected from the actual nature of the iface. What we really need (and have discussed in other PRs) is some kind of (possibly extensible) tagging system for ifaces that is independent of the L2, and somehow communicates the underlying nature of each iface. Then we could allow applications to ignore/unignore by tag. This would also fix the responsibility issue by making it the responsibility of the integration author to accurately tag the ifaces they provide, while still allowing the application developer full control over what tags they are interested in.

But the conclusion we have come to in those discussions is that such a tagging system is beyond the scope of this initial implementation.

I mainly included `net_conn_mgr_ignore_l2` specifically because of this sample -- it relies on filtering by L2 type, and that is not possible with the new event behaviors. If we had preserved the old event behavior and added new events with the new behavior, this would not be necessary -- but that was the approach taken in https://github.com/zephyrproject-rtos/zephyr/pull/53118, and that approach did not stand up to review

File: subsys/net/conn_mgr/conn_mgr.c
Comment (glarsennordic) [reply subsys/net/conn_mgr/conn_mgr.c]: Good catch! You're right, I chose to do that in a separate commit to help make reviewing easier. I forgot I had mentioned it in the commit message

File: subsys/net/conn_mgr/conn_mgr.c
Comment (glarsennordic) [reply subsys/net/conn_mgr/conn_mgr.c]: I'm not 100% sure which commit you commented on here, so just to clarify, you are talking about https://github.com/zephyrproject-rtos/zephyr/pull/56549/commits/02c02bb5de16a004c18b2c62ad2d8683faffc7a4 right?

Comment (glarsennordic) [reply subsys/net/conn_mgr/conn_mgr.c]: Whoops. I've had some troubles with propagating my event renames. Not sure how those changes keep ending up in later commits 🤦. Thanks for catching that, I'll get it corrected.

File: samples/net/sockets/echo_server/src/echo-server.c
Comment (glarsennordic) [reply samples/net/sockets/echo_server/src/echo-server.c:87]: I suppose these changes could be moved to a commit between this one and the commit before it -- But I included them in this commit since this is technically the commit that causes these changes to be required. Although in reality, the changes would work and are justifiable as early as hyr/pull/56549/commits/b8cf864359676657057323c35299cb533dbfdbc8

I'm pretty neutral, so final decision is up to you

Comment (rlubos) [reply samples/net/sockets/echo_server/src/echo-server.c:87]: Ah, after having a second look I now understand why are the changes related. I think that it could either be extracted into a separate commit after b8cf864, or, at very least, you could elaborate in the commit message on why are the changes needed.

File: samples/net/sockets/echo_server/src/tunnel.c
Comment (rlubos) [reply samples/net/sockets/echo_server/src/tunnel.c:171]: I start to wonder, is `SYS_INIT` configuration even needed here? IPIP interfaces have `NET_IF_NO_AUTO_START` set, so they won't start at boot, and we have `init_tunnel()` here which configures IP addresses, so this interface should not be considered "ready" before that. Especially, that the carrier state of an virtual interface is controlled by the operational UP status of the underlying interface - so the virtual iface should not be set to oper UP before the real one.

So I think the logic could be simplified a bit, and the interface could be ignored simply in `init_tunnel()`.

Comment (glarsennordic) [reply samples/net/sockets/echo_server/src/tunnel.c:171]: Ah! Good point! I will switch to the approach you suggest

File: samples/net/sockets/echo_server/src/echo-server.c
Comment (glarsennordic) [reply samples/net/sockets/echo_server/src/echo-server.c:87]: I'll do both, extract as a separate commit, and I'll try to update all the commit messages to include a bit more why, as per @jukkar 's very good point here: https://github.com/zephyrproject-rtos/zephyr/pull/57057#discussion_r1172540424

File: include/zephyr/net/conn_mgr.h
Comment (glarsennordic) [reply include/zephyr/net/conn_mgr.h:30]: I've improved the specificity of the documentation, although the behavior described technically requires a later commit to be included. I feel that is a lesser evil than rewriting the docs in the commit that changes the behavior. I will include a note about it in the commit message.

File: subsys/net/conn_mgr/conn_mgr_private.h
Comment (glarsennordic) [reply subsys/net/conn_mgr/conn_mgr_private.h:66]: Oops. I missed this before force-pushing. I'll nab it on the next revision.

File: include/zephyr/net/conn_mgr.h
Comment (glarsennordic) [reply include/zephyr/net/conn_mgr.h:56]: Same with this

**Review by rlubos — DISMISSED** [review]
Thanks, LGTM

File: include/zephyr/shell/shell_mqtt.h
Comment (glarsennordic) [reply include/zephyr/shell/shell_mqtt.h:12]: It's more of a subjective decision, none of us felt extremely strongly about it, but the placement feels more natural as opposed to it being in net/lib, since it is not really a library but a subsystem. I didn't include that reasoning in the commit since it is more subjective, especially with how Zephyr is organized in general. There isn't, to my knowledge, an objective argument for one location or another

**Comment by glarsennordic** [comment]
@jukkar I am having troubles running the tunneling overlay for echo_server, could you please advise? I see you are the one who added this overlay

I am able to run and ping the sample itself successfully in QEMU, but adding the tunnel overlay causes it to not get past boot.

I suspect it might be that I haven't configured `NET_SAMPLE_TUNNEL_PEER_ADDR` and `NET_SAMPLE_TUNNEL_MY_ADDR`, but I'm not actually sure what these should be set to. I might be able to figure that out, but I figure it makes sense to just, you know, ask the original author since you're right here 😅

**Comment by jukkar** [comment]
> I suspect it might be that I haven't configured NET_SAMPLE_TUNNEL_PEER_ADDR and NET_SAMPLE_TUNNEL_MY_ADDR, but I'm not actually sure what these should be set to. I might be able to figure that out, but I figure it makes sense to just, you know, ask the original author since you're right here

Sorry, I missed your question earlier. Not sure if the issue is still valid, but regarding the tunneling support, the capture API is perhaps the best example how it is being used. See for example `subsys/net/lib/capture` for the API and `subsys/net/ip/net_shell.c` (the `net capture` cmd) how the various parameters are set and what they mean. The `samples/net/capture` sample will enable the capture shell support if you want to try it live.

**Review by rlubos — DISMISSED** [review]

File: subsys/net/conn_mgr/events_handler.c
Comment (jukkar) [inline subsys/net/conn_mgr/events_handler.c:80]: Please use the `__fallthrough;` instead of the fallthrough comment.

File: subsys/net/conn_mgr/events_handler.c
Comment (jukkar) [inline subsys/net/conn_mgr/events_handler.c:86]: same here

**Comment by glarsennordic** [comment]
@jukkar I took a look at those samples, and they were helpful to some extent, but an actual test procedure is still unclear to me.

I have tried the following:

Terminal 1, in net-tools: `sudo ./net-setup.sh -c zeth-tunnel.conf`
Terminal 2, in net-tools: `./loop-socat.sh`
Terminal 3, in net-tools: `sudo ./loop-slip-tap.sh`
Terminal 4, in echo-server sample: `west build -p always -b qemu_x86 samples/net/sockets/echo_server -- -DCONF_FILE=overlay-tunnel.conf` followed by `west build -t run`

I derived this by trying to combine the instructions from these two sample pages:
https://docs.zephyrproject.org/3.2.0/samples/net/capture/README.html (per your recommendation)
https://docs.zephyrproject.org/2.7.0/guides/networking/qemu_setup.html (standard instructions for this sample)

Terminal 4 gets stuck after boot.

```
SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-20200714_234759-fv-az50-zephyr)
Booting from ROM..
*** Booting Zephyr OS build zephyr-v3.2.0-7691-ga01d40f6049a ***
```

This is true whether or not I use the latest from this PR, or 100 commits before it, suggesting that the issue is not with this PR.

My theory about improperly configured NET_SAMPLE_TUNNEL_PEER_ADDR and NET_SAMPLE_TUNNEL_MY_ADDR was wrong -- the overlay configures the exact values recommended by the files you pointed me to.

Do you have any recommendations on a proper test procedure for this sample with the tunnel overlay?

**Comment by rlubos** [comment]
Given that tunnel problem is reproducible on main, I don't think it shoudld block this PR. Please open a bug report regarding this sample configuration, I'll give it a closer look.

**Review by rlubos — DISMISSED** [review]

File: include/zephyr/net/conn_mgr.h
Comment (jukkar) [inline include/zephyr/net/conn_mgr.h:49]: This unignore word looks weird. I was not able to find it in dictionaries so it looks like it is not proper English. Dunno what we should have here, Perhaps `conn_mgr_start_ignore_iface()` and `conn_mgr_stop_ignore_iface()` although they are a bit longish.

File: subsys/net/conn_mgr/conn_mgr.c
Comment (jukkar) [inline subsys/net/conn_mgr/conn_mgr.c:50]: Why +1 here? It looks like you have now separate "index" space between this API and the net_if API.

File: include/zephyr/net/conn_mgr.h
Comment (glarsennordic) [reply include/zephyr/net/conn_mgr.h:49]: What about `conn_mgr_ignore_iface` and `conn_mgr_watch_iface`? This would align better with the documentation I wrote anyways. @rlubos @wbober You also brought up the naming of these two functions, would this naming be satisfying to you as well?

Comment (glarsennordic) [reply include/zephyr/net/conn_mgr.h]: I've gone ahead and force-pushed `ignore/unignore->ignore/watch` on the assumption/suspicion that everyone involved will either not care or prefer it. Let me know if this is satisfactory

File: subsys/net/conn_mgr/conn_mgr.c
Comment (jukkar) [reply subsys/net/conn_mgr/conn_mgr.c:50]: Some background for the index numbering. Originally network interface numbering started from 0, but then it was decided that index 0 would indicate all the network interfaces so similar was as in various Linux APIs that deal with network interfaces numbering. This means that net_if internally does +1 and -1 when figuring out the actual interface in the interface array.
So even if the index here is only used internally, I would suggest that we follow the index numbering style done in net_if. It is just quite confusing otherwise for the future developer who is looking the code and trying to understand why the index numbering differs. Also this could avoid possible hard to find bugs.

File: include/zephyr/net/conn_mgr.h
Comment (jukkar) [reply include/zephyr/net/conn_mgr.h]: Ignore/watch naming LGTM.

Comment (rlubos) [reply include/zephyr/net/conn_mgr.h]: Sounds fine to me

**Comment by glarsennordic** [comment]
If there are no further concerns, would you be willing to approve, @jukkar @rlubos ?

**Review by rlubos — APPROVED** [review]

File: subsys/net/conn_mgr/conn_mgr.c
Comment (glarsennordic) [reply subsys/net/conn_mgr/conn_mgr.c:54]: I have now done as you suggest, Robert! @jukkar let me know if this change is satisfactory

File: subsys/net/conn_mgr/conn_mgr.c
Comment (jukkar) [inline subsys/net/conn_mgr/conn_mgr.c:49]: I would have thought that the index here is the interface index number, and not the iface[index]. This is what I mean that using the same name for this value is super confusing. Especially if the documentation does not clearly indicate what the value is like here. The iface_states is documented but imho documentation here would need some enhancements.

File: subsys/net/conn_mgr/conn_mgr.c
Comment (jukkar) [inline subsys/net/conn_mgr/conn_mgr.c:27]: I am just wondering why the indexing cannot start from 1 for the iface_states too. In the getter/setter function you would then just do iface_states[index - 1]. Requires of course that one does not access the iface_states directly but via a function.

File: subsys/net/conn_mgr/conn_mgr.c
Comment (rlubos) [reply subsys/net/conn_mgr/conn_mgr.c:49]: Isn't the following part explicit enough that the indexing refers to `iface_states` array?
> by the index that corresponds to it in **iface_states**

I can't really see an alternative for this function. And for details on what `iface_states` is, it has it's own documentation, I'm not sure if repeating the information here is beneficial.

File: subsys/net/conn_mgr/conn_mgr.c
Comment (rlubos) [reply subsys/net/conn_mgr/conn_mgr.c:27]: IMO it could then can confusing in other place, where we simply iterate over the array - it be needed then not to use C-style array indexing (from 0) but start from 1 and use getter/setter. Perhaps it's just a matter of preference, but I don't think the current approach is that confusing (we no longer store the modified index in `last_iface_down/up`).

**Review by rlubos — APPROVED** [review]

File: subsys/net/conn_mgr/conn_mgr.c
Comment (jukkar) [reply subsys/net/conn_mgr/conn_mgr.c:27]: Fair enough, if you think this is clear enough I can certainly live with this.

**Review by jukkar — APPROVED** [review]

**Pull request merged**
